File: 문제풀이/온라인 저지/백준/# 05. Bronze Ⅰ/# 09093. 단어 뒤집기/java/Main.java
import java.io.DataInputStream;
import java.io.DataOutputStream;

class Stack {
    private int size;
    private final byte[] arr = new byte[20];

    void push(byte b) {
        arr[size++] = b;
    }

    byte pop() {
        return arr[--size];
    }

    boolean isNotEmpty() {
        return size != 0;
    }
}

public class Main {

    public static void main(String[] args) {
        int t = i();
        Stack stack = new Stack();
        byte c, sep;
        while (t > 0) {
            sep = ASCII_space;
            while (true) {
                c = r();
                if (c == ASCII_n) {
                    t --;
                    sep = c;
                }
                if (c <= ASCII_space) break;
                stack.push(c);
            }
            while (stack.isNotEmpty()) writeByte(stack.pop());
            writeByte(sep);
        }
        flushBuffer();
    }

    private static final byte EOF = -1;
    private static final byte ASCII_n = 10;
    private static final byte ASCII_space = 32;
    private static final byte ASCII_0 = 48;
    private static final DataInputStream din = new DataInputStream(System.in);
    private static final DataOutputStream dout = new DataOutputStream(System.out);
    private static final byte[] inbuffer = new byte[65536];
    private static int inbufferpointer;
    private static int bytesread;
    private static final byte[] outbuffer = new byte[65536];
    private static int outbufferpointer;

    private static int i() {
        int v = 0;
        byte c = r();
        do {v = v * 10 + c - 48;} while (isDigit((c = r())));
        return v;
    }

    private static byte r() {
        if (inbufferpointer == bytesread) fillBuffer();
        return (bytesread == EOF) ? EOF : inbuffer[inbufferpointer++];
    }

    private static void fillBuffer() {
        try {
            bytesread = din.read(inbuffer, (inbufferpointer = 0), inbuffer.length);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }

    private static void writeByte(byte b) {
        if (outbufferpointer == outbuffer.length) flushBuffer();
        outbuffer[outbufferpointer++] = b;
    }

    private static void flushBuffer() {
        if (outbufferpointer != 0) {
            try {
                dout.write(outbuffer, 0, outbufferpointer);
            } catch (Exception e) {
                throw new RuntimeException(e);
            }
            outbufferpointer = 0;
        }
    }

    private static boolean isDigit(byte b) {
        return b >= ASCII_0;
    }

}

File: 문제풀이/온라인 저지/백준/# 05. Bronze Ⅰ/# 09093. 단어 뒤집기/kotlin/Main.kt
fun main() {
    var t = i()
    val stack = Stack()

    fun s() {
        var c = r()
        var sep = ASCII_space
        do stack.push(c) while (!isSpace(r().also { c = it; if (isN(c)) {t--; sep = c} }))
        while(stack.isNotEmpty()) writeByte(stack.pop())
        writeByte(sep)
    }
    while(t > 0) s()
    flushBuffer()
}

class Stack {

    private var size = 0
    private val arr = ByteArray(20)

    fun push(b: Byte) {
        arr[size++] = b
    }

    fun pop() = arr[--size]

    fun isNotEmpty() = size != 0
}

private const val EOF = -1
private const val ASCII_n: Byte = 10
private const val ASCII_space: Byte = 32
private const val ASCII_0: Byte = 48
private val din = java.io.DataInputStream(System.`in`)
private val dout = java.io.DataOutputStream(System.out)
private val inbuffer = ByteArray(65536)
private var inbufferpointer = 0
private var bytesread = 0
private val outbuffer = ByteArray(65536)
private var outbufferpointer = 0

private fun i(): Int {
    var v = 0
    var c = r()
    do v = v * 10 + c - 48 while (isDigit(r().also { c = it }))
    return v
}

private fun r(): Byte {
    if (inbufferpointer == bytesread) fillBuffer()
    return if (bytesread == EOF) EOF.toByte() else inbuffer[inbufferpointer++]
}

private fun fillBuffer() {
    bytesread = din.read(inbuffer, 0.also { inbufferpointer = it }, inbuffer.size)
}

private fun writeByte(b: Byte) {
    if (outbufferpointer == outbuffer.size) flushBuffer()
    outbuffer[outbufferpointer++] = b
}

private fun flushBuffer() {
    if (outbufferpointer != 0) {
        dout.write(outbuffer, 0, outbufferpointer)
        outbufferpointer = 0
    }
}

private fun isSpace(b: Byte) = b <= ASCII_space
private fun isN(b: Byte) = b <= ASCII_n
private fun isDigit(b: Byte) = b >= ASCII_0
